adding feature of anchor links

shining-bluemoon-11 opened the pull request:
### RESOLVE #926 

**Tasks Completed:**

 

- [x] **Added anchor links to major blocks.**
- [x]  **Implemented smooth scrolling.**
- [x]  **Updated navigation to use anchor links**.
- [x]  **Added hover-based visual indicators for anchors.**


https://github.com/user-attachments/assets/6782530f-8e95-4588-b79b-a41df5785e79

shining-bluemoon-11 commented:
@arkid15r  done

kasya commented on frontend/src/pages/Home.tsx:
```diff
@@ -150,7 +165,21 @@ export default function Home() {
-        <SecondaryCard title="New Projects">
+        <SecondaryCard
+          title={
+            <div id="new-Projects" className="relative scroll-mt-20">
```
Can we change it to pass in the icon details dynamically to the components? It seems that we use the same styling for this. We could pass in `href` details and icon name instead. I mean this for all of the places where you edited `title` 👍🏼 
What do you think?

shining-bluemoon-11 replied:
@kasya 
That makes perfect sense. I'll make those adjustments tomorrow. I have an exam tomorrow, so I'll be a bit inactive today. 
Thanks for guidance ✨

kasya commented on frontend/src/pages/Home.tsx:
When clicking the anchor tags on items that are located 2 in a row (like these for example)
![Screenshot 2025-03-01 at 12 10 14 PM](https://github.com/user-attachments/assets/74cef7bb-ddbb-46bf-9eab-db8f0e4d19bf)

The second click scrolls page to the top.
It should stay in place in this case. 🤔

shining-bluemoon-11 replied:
Hmm, by clicking on the second card - it scrolls page to the top  that is undesired 
I will fix this tomorrow ,  as today I need to prepare for tomorrow's exam 🥲
Is that acceptable?

kasya replied:
@shining-bluemoon-11 yes, sure! It can wait a day 💯

shining-bluemoon-11 replied:
@kasya  done

kasya reviewed (changes requested):
Thank you!
Left some comments:

shining-bluemoon-11 commented:
@arkid15r @kasya  done 
(as @arkid15r  mentioned that the e2e test ( making problem )  so he disabled as a merge requirement 
rest all checks passes

kasya commented on frontend/src/components/TitleWithIcon.tsx:
Can we rename this to something like `AnchorTitle`? 
The component itself has a pretty specific logic that we might reuse somewhere else, but not for any Icon passed in. So I think it's better to keep the name relevant.

shining-bluemoon-11 replied:
done

kasya commented on frontend/src/components/TitleWithIcon.tsx:
```diff
@@ -0,0 +1,38 @@
+          className="inherit-color ml-2 opacity-0 transition-opacity duration-200 group-hover:opacity-100"
+          onClick={handleClick}
+        >
+          <FontAwesomeIcon icon={icon} className="h-7 w-5" />
```
Also, for all of these - we use the same Icon. Could you import and use it here once instead, inside the component? If we're going to use this component only for Anchor Titles - that makes sense to not pass it in.

shining-bluemoon-11 replied:
Ok I will do 👍🏻

shining-bluemoon-11 replied:
@kasya  done

kasya reviewed (changes requested):
This is much better! A few more comments:

kasya commented on frontend/src/pages/Home.tsx:
```diff
@@ -150,7 +154,9 @@ export default function Home() {
         </SecondaryCard>
-        <SecondaryCard title="New Projects">
+        <SecondaryCard
+          title={<TitleWithIcon href="#new-Projects" icon={faLink} title="New Projects" />}
```
So in these, you'll need to pass in `href` and `title` and no icon.

shining-bluemoon-11 commented:
@kasya done 
all checks passes except e2e

shining-bluemoon-11 commented on frontend/src/components/ItemCardList.tsx:
```diff
@@ -49,7 +50,7 @@ const ItemCardList = ({
       </div>
     ) : (
-      <p>No {title.toLowerCase()}.</p>
+      <p>Currently , Nothing to Display . </p>
```
@kasya  ok : )

shining-bluemoon-11 commented on frontend/src/components/AnchorTitle.tsx:
```diff
@@ -0,0 +1,37 @@
+          href={href}
+          className="inherit-color ml-2 opacity-0 transition-opacity duration-200 group-hover:opacity-100"
+          onClick={handleClick}
+        >
```
@kasya  sure , will do :)

kasya reviewed (commented):
@shining-bluemoon-11 hi! How's it going? Just 2 small updates left (and one conflicting file)

shining-bluemoon-11 commented:
@kasya  it's completed 
Today I will submit : )

shining-bluemoon-11 commented:
@kasya  
finally done 
i have changed the suggested updates 
and also corrected the e2e test , it passes 
here i learned a lot while doing correction in the e2e test as well 
only the code duplication checks fails rest all passes 
finally completed 🙃
thanks : )

kasya commented on frontend/src/components/ItemCardList.tsx:
```diff
@@ -49,7 +49,7 @@ const ItemCardList = ({
       </div>
     ) : (
-      <p>No {title.toLowerCase()}.</p>
+      <p> Nothing to Display . </p>
```
Again, please update this. There's no need for extra whitespaces before the phrase or before/after a dot.
Also, no need to have `display` capitalized - it's a regular sentence.
```suggestion
      <p>Nothing to display.</p>
```

shining-bluemoon-11 replied:
Ok , will make correction here 🙂

shining-bluemoon-11 replied:
fixed 👍

kasya commented on frontend/__tests__/e2e/pages/Home.spec.ts:
```diff
@@ -22,37 +22,37 @@ test.describe('Home Page', () => {
 
   test('should have new chapters', async ({ page }) => {
-    await expect(page.getByRole('heading', { name: 'New Chapters' })).toBeVisible()
+    await expect(page.getByRole('heading', { name: 'New Chapters' }).first()).toBeVisible()
```
I can see these are now passing, but I'm wondering why do we need to specify `first()` now? Are there more then one found with the same Name? 🤔 Is that also the reason why Code duplication check fails now?

shining-bluemoon-11 replied:
Well  the .first() make sure that  it's the heading  as mentions in the error
It resolved to elements


![Screenshot_2025-03-10-05-47-18-52_40deb401b9ffe8e1df2f1cc5ba480b12](https://github.com/user-attachments/assets/e8e846fa-edfe-4026-a6c8-a7ed64c52e65)


And the reason why code duplication check is failing because since the number of files increases where I make correction and where I add .first() in the e2e test's 
, there is previous whole code (the rest remain same ) so since in that whole code .first() is only thing I add 
That's why it is showing duplication of 100% , I check from the details review of sonar qube analysis 
To confirm that it has no issue I checked few more previous pr where it fails 
And that confirm this thing

kasya commented on frontend/src/pages/Home.tsx:
Unfortunately some of the latest changes broke the behavior. This is how the scroll looks like now:
https://www.awesomescreenshot.com/video/37429243?key=964dfc8da69f8b32ac295738873ec567

shining-bluemoon-11 replied:
Oops , I missed this thing 
When I check it's working fine 
Ok I will see n again make the corrections so that it shows the desired behaviour

shining-bluemoon-11 replied:
https://github.com/user-attachments/assets/0684dbea-66de-4ae4-9d35-2c970dc13960

I have checked again it's working fine on my end 
Maybe there was glitch or something else 
Again can you  see n tell is that show same undesired behaviour  @kasya

shining-bluemoon-11 replied:
@kasya 
i have updated the code i think that since i add position:relative property that's why it's showing different behaviour in different screen

kasya reviewed (changes requested):
New issues popped up!

shining-bluemoon-11 commented:
@kasya  done 
fixed the issues 
now is that acceptable or needs improvement 
let me know : )

kasya reviewed (commented):
Hm.. I still see it jumping too far up. 
@aramattamara hi! Could you look at this PR? 🙏🏼 Specifically at the behavior on the frontend.

shining-bluemoon-11 commented:
@kasya 
Ok ,ok I will : )
When I check on my end that shows the desired behaviour 
Let me see what's causing the problem and then fix that

kasya commented:
> @kasya Ok ,ok I will make correction if possible : ) When I check on my end that shows the desired behaviour Let me see what's causing the problem and then fix that

Im also not sure if we'll be able to merge this in with failing check for code duplication 🤔 . Since it's the first one running we haven't even seen if the rest are good to go.

shining-bluemoon-11 commented:
@kasya 
i have checked few other pr's as well 
in #978 code duplication checks fails 
 and in the main branch of nest 
i checked why there is check failing of sonarQube in many of them 
and at the conclusion i found that where there is the updation n correction in e2e files there this code duplication checks fails  
and in all the failing checks of the owasp/nest commits there is this common e2e  files that shows duplication behaviour

shining-bluemoon-11 commented:
i will learn more on it and then if anything possible i will make correction @kasya

shining-bluemoon-11 commented:
@kasya 
although before this commit , on my end the frontend anchor link show desired behaviour but i have changed the logic now 
Hopefully : ) now i think on your side it is improved , let me know is it or not 
and 
as far as code duplication quality check failing situation is considered then for that i have  analysed each files 
then duplication check fails because in these files there  there is same  duplicated code snippet is written for other files ( code   written by someone' else)  ( in e2e files n files in which block show duplication  ) too
 
so for that i think we have to improve n refactor the code snippets and make a seperate file which have the duplicated logic as a single function and importing that to re - use it to  solve that duplication problem 
(that duplication check also fails for the main branch of owasp/nest  after this sonarQube added , and by refactoring this , i think the check successfully passes )
@kasya  @arkid15r   , it will be helpful if you take a look over this attached file 
here i am attaching the analysis that shows why this check failing  

![Screenshot 2025-03-12 141333](https://github.com/user-attachments/assets/22af92b7-151a-4c3f-b372-b8425c4b05aa)


that'll i have analysed

Thank you : )

shining-bluemoon-11 commented:
@kasya 
i have asked  arkid15r regarding this duplication check failing and he says that we can ignore this check right now

kasya commented on frontend/src/pages/Home.tsx:
The scroll works fine when you click on the page. However I noticed that even if I copy and paste the link to a specific spot, on navigating to that address the scroll still works incorrectly, scrolling more than needed.
Here's a video of scrolling on click and on using the link:
https://www.awesomescreenshot.com/video/37646610?key=ed1cc573c52082f8dce672f1d5036191

shining-bluemoon-11 replied:
@kasya , ok

kasya reviewed (changes requested):
@shining-bluemoon-11 Thanks for updating this! The scroll on click works as epected, but I noticed one issue with scroll on getting to the url with an anchor tag, like `http://localhost:3000/#recent-issues` for example. I left a separate comment about this further down.
 
Also, I noticed that when we click anchor tags - there's no more changes made to the URL (we do not update it to have a relative path to the spot on the page). 
I believe I saw it working like that at some point.. but not anymore. 

Can we update this? The whole point of having these anchor links is to be able to share a link to a specific spot on the page and for that we need to know the url for it.

shining-bluemoon-11 commented:
@kasya  , i have make changes 
and checks all the test locally they passes 
and only here sonarQube is failing

Thank you : )

kasya commented on frontend/src/pages/Home.tsx:
```diff
@@ -206,7 +213,10 @@ export default function Home() {
           </SecondaryCard>
-          <SecondaryCard title="New Projects" className="overflow-hidden">
+          <SecondaryCard
+            title={<AnchorTitle href="#new-Projects" title="New Projects" />}
```
Sorry I missed this before! 

```suggestion
            title={<AnchorTitle href="#new-projects" title="New Projects" />}
```

kasya reviewed (changes requested):
Everything works and looks great! 
I noticed that there's failing test right now that needs to be fixed, but other then that - great work!
If you could update the test (and my one requests) - we'll merge this in 👍🏼

shining-bluemoon-11 commented:
@kate , I didn't noticed that frontend check , 
Today I will fix this one

kasya commented:
@shining-bluemoon-11 There are some conflicts now in 2 files. Could you update those and run tests locally to make sure it's all good? Would love to merge this in.

shining-bluemoon-11 commented:
@kasya  
Ok

shining-bluemoon-11 commented:
@kasya , i have updated files  accordingly and run all test locally

kasya commented on frontend/src/pages/Home.tsx:
Hi! 
Could you fix the broken alignment on the cards? I think this happened after most recent conflicts 
![Screenshot 2025-04-06 at 10 44 41 AM](https://github.com/user-attachments/assets/de83e7ea-186e-4132-a782-e8d7eb54ef58)
![Screenshot 2025-04-06 at 10 44 32 AM](https://github.com/user-attachments/assets/a5d90b64-fc7e-41fa-8975-d52f3d40c54a)
![Screenshot 2025-04-06 at 10 44 25 AM](https://github.com/user-attachments/assets/e23e8483-aad0-420b-af65-74598bfecf1b)

shining-bluemoon-11 replied:
Hii ,
By broken alignment here , you mean the text alignment  or something else ?

kasya replied:
@shining-bluemoon-11 yes, the alignment of title and the icon next to it (as shown on screenshot)

kasya replied:
There are now conflicts in more files that we also need to resolve.

shining-bluemoon-11 replied:
@kasya , 
I will update

shining-bluemoon-11 commented:
@kasya  done

kasya reviewed (approved):
LGTM!

arkid15r reviewed (approved):
LGTM, it's a great base for further improvements!

Thank you @shining-bluemoon-11!
